### PR TITLE
Sparse file support (integration)

### LINKED ISSUE #5566: chunker: sparsify if all-zero data was read
related to #14: 

once we have better sparse handling inside borg (communicating data vs. hole status between chunker, hashing, ...), it makes sense to add a sparsify functionality.

assume we read all-zero chunks from a file:

1. could be a hole (no space allocated on disk) [we can avoid this with upcoming change from #5561]
2. could be file space allocated, but never used (fallocate)
3. could be zeros stored on disk

for the last 2 cases, we could add a `--sparsify` to `borg create`:

just compare the chunk content to an all-zero bytearray.

if it compares **equal** and `--sparsify` was given, the chunker shall (at least) yield "this is a hole of length X".

later it could yield (if we can detect somehow):
- "this is a not-allocated hole of length X" (case 1)
- "this is fallocated unused space of length X" (case 2)
- "this is on-disk all-zero space of length X" (case 3)

### PR DESCRIPTION
**Pull request by ThomasWaldmann**
fixes #5566 

the chunker now yields namedtuples (data, meta), so that sparse holes can be processed differently than normal data.

to avoid hashing the all-zero blocks again and again, a LRUcache for a few popular block sizes and their hashes is maintained.
this speeds up processing of sparse holes / all-zero blocks a lot.

additionally to all-zero sparse holes, the fixed-blocksize chunker now can also detect all-zero from-disk reads (allocation type CH_ALLOC).

also refactored all places that need to compare against `zeros` so that we only have one big all-zero-bytes object.

also added all-zeros detection for the buzhash chunker (here the results for a huge, completely sparse file).

Some stats:
- buzhash chunker, before this PR: Duration: 2 minutes 20.14 seconds
- buzhash chunker, after this PR: Duration: 1 minutes 2.62 seconds
- fixed chunker, before this PR: Duration: 1 minutes 25.48 seconds
- fixed chunker, without --sparse, after this PR: Duration: 11.48 seconds
- fixed chunker, with --sparse, after this PR: Duration: 0.42 seconds

### CODE REVIEW

File: src/borg/testsuite/chunker.py [hunk 1]
```diff
@@ -8,18 +8,32 @@
...
+    """chunk filter"""
+    # this is to simplify testing: either return the data piece (bytes) or the hole length (int).
+    def _cf(chunk):
+        if chunk.meta['allocation'] == CH_DATA:
+            assert len(chunk.data) == chunk.meta['size']
+            return bytes(chunk.data)  # make sure we have bytes, not memoryview
+        if chunk.meta['allocation'] == CH_HOLE:
+            assert chunk.data is None
+            return chunk.meta['size']
+        assert False, "unexpected allocation value"
```
Comment (ThomasWaldmann) [inline src/borg/testsuite/chunker.py:21]: CH_ALLOC?

File: src/borg/chunker.pyx [hunk 2]
```diff
@@ -26,6 +29,25 @@ cdef extern from "_chunker.c":
...
+
+    meta is always a dictionary, data depends on allocation.
+
+    on disk data:
+        meta = {'allocation' = CH_DATA, 'size' = size_of_data }
+        data = read_data [bytes or memoryview]
+
+    hole in a sparse file:
+        meta = {'allocation' = CH_HOLE, 'size' = size_of_hole }
+        data = None
```
[the remark this answers is not in the document]
Comment (ThomasWaldmann) [reply src/borg/chunker.pyx:44]: fixed

File: src/borg/archive.py [hunk 3]
```diff
@@ -336,7 +337,7 @@ def flush(self, flush=False):
         self.buffer.seek(0)
         # The chunker returns a memoryview to its internal buffer,
         # thus a copy is needed before resuming the chunker iterator.
-        chunks = list(bytes(s) for s in self.chunker.chunkify(self.buffer))
+        chunks = list(bytes(chunk.data) for chunk in self.chunker.chunkify(self.buffer))
```
[the remark this answers is not in the document]
Comment (ThomasWaldmann) [reply src/borg/archive.py:340]: added a comment in the src.

File: src/borg/archive.py [hunk 4]
```diff
@@ -1982,7 +1999,10 @@ def process_chunks(self, archive, target, item):
         chunk_processor = partial(self.chunk_processor, target)
         target.process_file_chunks(item, self.cache, target.stats, self.progress, chunk_iterator, chunk_processor)
 
-    def chunk_processor(self, target, data):
+    def chunk_processor(self, target, chunk):
+        # as this is recreate (we do not read from the fs), we never have holes here
+        assert chunk.meta['allocation'] == CH_DATA
```
[the remark this answers is not in the document]
Comment (ThomasWaldmann) [reply src/borg/archive.py:2004]: todo added

File: src/borg/archive.py [hunk 5]
```diff
@@ -1102,6 +1103,8 @@ def __init__(self, *, key, cache,
         self.checkpoint_interval = checkpoint_interval
         self.last_checkpoint = time.monotonic()
         self.rechunkify = rechunkify
+        self.zero_chunk_ids = LRUCache(10, dispose=lambda _: None)  # length of all-zero chunk -> chunk_id
+        self.zeros = memoryview(bytes(MAX_DATA_SIZE))
```
[the remark this answers is not in the document]
Comment (ThomasWaldmann) [reply src/borg/archive.py]: done

File: src/borg/chunker.pyx [hunk 6]
```diff
@@ -178,15 +200,20 @@ class ChunkerFixed:
...
-                    data = self.zeros[:pos - offset]  # for now, create zero-bytes here
-                got = len(data)
+                    got = pos - offset
+                    data = None
+                    is_zero = True
                 if got > 0:
                     offset += got
                     range_size -= got
-                    yield data  # later, use a better api that tags data vs. hole
+                    yield Chunk(data, size=got, allocation=(CH_ALLOC if is_zero else CH_DATA) if is_data else CH_HOLE)
```
[the remark this answers is not in the document]
Comment (ThomasWaldmann) [reply src/borg/chunker.pyx]: fixed

File: src/borg/testsuite/chunker.py (hunk 1 above)
Comment (ThomasWaldmann) [reply src/borg/testsuite/chunker.py]: fixed

**Review by timmc — APPROVED** [review]
Nice. One nitpick on some maybe overly dense code, but nothing important.

Is there anything preventing the content-defined chunker from detecting all-zeroes chunks? I expect with that implementation, almost all zero-chunks would be the same size as each other (the max size param).

File: src/borg/chunker.pyx [hunk 7]
```diff
@@ -178,15 +204,22 @@ class ChunkerFixed:
...
```
[the remark this answers is not in the document]
Comment (ThomasWaldmann) [reply src/borg/chunker.pyx]: fixed :-)

**Comment by ThomasWaldmann** [comment]
About "Is there anything preventing the content-defined chunker from detecting all-zeroes chunks? I expect with that implementation, almost all zero-chunks would be the same size as each other (the max size param)."

Now you say it... - this can be easily done, see latest commit. I just did not want to add real SEEK_DATA/SEEK_HOLE support to the C code. But detecting if we got all-zero is easy.

File: src/borg/chunker.pyx (hunk 7 above)
[the remark this answers is not in the document]
Comment (timmc) [reply src/borg/chunker.pyx]: Cool, that actually cleans up better than I was imagining.

**Pull request merged**